analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/NullPointerDereferenceBase.cs

```diff
@@ -20,34 +20,45 @@
 
-        protected override ProgramState PreProcessSimple(SymbolicContext context)
+    protected override ProgramState PreProcessSimple(SymbolicContext context)
+    {
+        if (ReferenceOrDefault(context.Operation.Instance) is { Syntax: var syntax} reference
+            && context.HasConstraint(reference, ObjectConstraint.Null)
+            && !IsSupressed(syntax)
+            && SemanticModel.GetTypeInfo(syntax).Nullability().FlowState != NullableFlowState.NotNull)
         {
-            var reference = context.Operation.Instance.Kind switch
-            {
-                OperationKindEx.Invocation => context.Operation.Instance.ToInvocation().Instance,
-                OperationKindEx.PropertyReference => context.Operation.Instance.ToPropertyReference().Instance,
-                OperationKindEx.Await => context.Operation.Instance.ToAwait().Operation,
-                OperationKindEx.ArrayElementReference => context.Operation.Instance.ToArrayElementReference().ArrayReference,
-                _ => null,
-            };
-            if (reference != null
-                && context.HasConstraint(reference, ObjectConstraint.Null)
-                && !reference.Type.IsStruct() // ToDo: IsStruct() is a workaround before MMF-2401
-                && !IsSupressed(reference.Syntax)
-                && SemanticModel.GetTypeInfo(reference.Syntax).Nullability().FlowState != NullableFlowState.NotNull)
-            {
-                ReportIssue(reference, reference.Syntax.ToString());
-            }
-
-            return context.State;
+            ReportIssue(reference, syntax.ToString());
         }
+        return context.State;
     }
+
+    private static IOperation ReferenceOrDefault(IOperation operation) =>
+        operation.Kind switch
+        {
+            OperationKindEx.Invocation => InvocationOrDefault(operation),
+            OperationKindEx.PropertyReference => PropertyReferenceOrDefault(operation),
+            OperationKindEx.Await => operation.ToAwait().Operation,
+            OperationKindEx.ArrayElementReference => operation.ToArrayElementReference().ArrayReference,
+            _ => null,
+        };
+
+    private static IOperation InvocationOrDefault(IOperation operation) =>
+        operation.ToInvocation() is { TargetMethod: var method } invocation
+        && !method.IsAny(KnownType.System_Nullable_T, nameof(Nullable<int>.GetValueOrDefault), nameof(Nullable<int>.Equals), nameof(Nullable<int>.ToString), nameof(Nullable<int>.GetHashCode))     // GetType raises, since it throws NRE
+            ? invocation.Instance
+            : null;
+
+    private static IOperation PropertyReferenceOrDefault(IOperation operation) =>
+        operation.ToPropertyReference() is { Property: var property } propertyReference
+        && !(property.IsInType(KnownType.System_Nullable_T) && property.Name is nameof(Nullable<int>.HasValue) or nameof(Nullable<int>.Value))  // HasValue doesn't throw; Value is covered by S3655
```

Comment:
Good catch. I prefer checking it for the following reasons, though:
1. The Intend is much clearer.
2. If more properties get added in the future, it will err on the side of FN instead of FP.

From what I can tell, checking the name should not have a significant performance cost, so I would prefer keeping it.